src/main/common/vector.c

```diff
@@ -0,0 +1,174 @@
+/*
+ * This file is part of Cleanflight and Betaflight.
+ *
+ * Cleanflight and Betaflight are free software. You can redistribute
+ * this software and/or modify this software under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+
+#include <math.h>
+
+#include "common/axis.h"
+#include "common/maths.h"
+
+#include "vector.h"
+
+void vector2Zero(vector2_t *v)
+{
+    v->x = 0.0f;
+    v->y = 0.0f;
+}
+
+void vector2Add(vector2_t *result, const vector2_t *a, const vector2_t *b)
+{
+    result->x = a->x + b->x;
+    result->y = a->y + b->y;
+}
+
+void vector2Scale(vector2_t *result, const vector2_t *a, const float k)
+{
+    result->x = a->x * k;
+    result->y = a->y * k;
+}
+
+float vector2Dot(const vector2_t *a, const vector2_t *b)
+{
+    return a->x * b->x + a->y * b->y;
+}
+
+float vector2Cross(const vector2_t *a, const vector2_t *b)
+{
+    return a->x * b->y - a->y * b->x;
+}
+
+float vector2NormSq(const vector2_t *v)
+{
+    return vector2Dot(v, v);
+}
+
+float vector2Norm(const vector2_t *v)
+{
+    return sqrtf(vector2NormSq(v));
+}
+
+void vector2Normalize(vector2_t *result, const vector2_t *v)
+{
+    float normSq = vector2NormSq(v);
+    if (normSq > 0.0f) {
+        vector2Scale(result, v, 1.0f / sqrtf(normSq));
+    } else {
+        vector2Zero(result);
+    }
+}
+
+void vector3Zero(vector3_t *v)
+{
+    v->x = 0.0f;
+    v->y = 0.0f;
+    v->z = 0.0f;
+}
+
+void vector3Add(vector3_t *result, const vector3_t *a, const vector3_t *b)
+{
+    result->x = a->x + b->x;
+    result->y = a->y + b->y;
+    result->z = a->z + b->z;
+}
+
+void vector3Scale(vector3_t *result, const vector3_t *a, const float k)
+{
+    result->x = a->x * k;
+    result->y = a->y * k;
+    result->z = a->z * k;
+}
+
+float vector3Dot(const vector3_t *a, const vector3_t *b)
+{
+    return a->x * b->x + a->y * b->y + a->z * b->z;
+}
+
+void vector3Cross(vector3_t *result, const vector3_t *a, const vector3_t *b)
+{
+    result->x = a->y * b->z - a->z * b->y;
```

Comment:
I made this on purpose for efficiency and to prevent unneeded copy instructions. But now everyone using this function needs to go into a contract, where `result` needs to differ from `a` and `b` at all times. This is a big potential tripping hazard, I totally agree with you...

Sacrificing a bit of efficiency for more resilient behavior is probably the right call here, thanks. I'll change it back to use a temporary copy of each vector.